Feat: add access request & granting

NuttyShrimp opened the pull request:
Adds access request & inbox tabs to handle:
- requesting certain permissions from requestable files
- Visualizing incoming request & logic to accept or reject these request
- An inbox with the response on the request to get certain permissions

## How to test
I've used firefox built-in container tabs to prevent the need of 2 dev server running in the background.

closes #9

NuttyShrimp committed:
feat(header): move header to layout & add new request page

NuttyShrimp committed:
feat(controller): add controls for resources.json

NuttyShrimp committed:
feat(loama): add toggle to allow access request for a resource

NuttyShrimp committed:
feat(loama): add url to set webID for request access to pod resources

NuttyShrimp committed:
refactor(controller): make store more abstract for re-usability sake

NuttyShrimp committed:
refactor(controller): move request logic to seperate class

NuttyShrimp committed:
fix: update loama to match changes of previous commit

NuttyShrimp committed:
feat: get data as tree + transform into usable data for library

NuttyShrimp committed:
fix(lobutton): fix inconsistency in border setting

NuttyShrimp committed:
feat(accessRequest): add check if inbox exists when pod url is changed

NuttyShrimp committed:
refactor(controller): move store creation into controller & modules

NuttyShrimp committed:
feat(accessRequest): add inbox store

NuttyShrimp committed:
fix(store): support resource in containers

NuttyShrimp committed:
feat(accessRequest): send accessRequest to loama inbox

NuttyShrimp committed:
feat(accessRequest): use acl link from resource info in request

NuttyShrimp committed:
fix(accessRequest): also make sure resources.json is created before setting permissions

NuttyShrimp committed:
refactor(inbox): add inbox specific type & constructor

NuttyShrimp committed:
fix(accessRequest): filter requested resource to prevent requests for non-requestable resources

NuttyShrimp committed:
fix(accessRequest): make directories requestable

NuttyShrimp committed:
fix(button): also add colormix to button border

NuttyShrimp committed:
feat(loama): add inbox view to visualize access requests

NuttyShrimp committed:
feat(accessRequest): introduce response logic to access request

NuttyShrimp committed:
feat(accessRequest): send more data in request response

NuttyShrimp committed:
fix(inrupAccessRequest): await file removal

NuttyShrimp committed:
feat(controller): Add logic to retrieve access request responses

NuttyShrimp committed:
feat(loama): add inbox to view & remove access request responses

NuttyShrimp committed:
fix(controller): fix rebase issues

NuttyShrimp committed:
fix(controller): await setpodurl

Makes that the new files are directly available & have their
corresponding permissions

elsdvlee reviewed (commented):
I have not looked at the code, but intended to test the functionality. 
Doing this, I ran into a first issue: it is unclear to me as a user how to request access. 
What is expected in the field WebID: the WebID to whom I want to sent an access request? 
What is expected in the field next to it? 
Where do I add the path of the file to which I request access? 
![image](https://github.com/user-attachments/assets/f4864ceb-8c7e-4a37-b326-8690ae1cddcb)
See screenshot: I logged in as [electronics_user1](https://onto-deside.ilabt.imec.be/css11/electronics_user1/profile/card#me). The button 'request access' remains disabled and the field above it as well.

NuttyShrimp committed:
feat(requestView): add extra information to inputs

NuttyShrimp commented:
> I have not looked at the code, but intended to test the functionality. Doing this, I ran into a first issue: it is unclear to me as a user how to request access. What is expected in the field WebID: the WebID to whom I want to sent an access request? What is expected in the field next to it? Where do I add the path of the file to which I request access? ![image](https://private-user-images.githubusercontent.com/93709574/389487308-f4864ceb-8c7e-4a37-b326-8690ae1cddcb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI1NDcwODAsIm5iZiI6MTczMjU0Njc4MCwicGF0aCI6Ii85MzcwOTU3NC8zODk0ODczMDgtZjQ4NjRjZWItOGM3ZS00YTM3LWIzMjYtODY5MGFlMWNkZGNiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI1VDE0NTk0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThiMzcwNmUwZGI5MGMyZDBhMGUyYjQ4YmFlYzllNzE1MDE3MzhlODA3OGVmZGQ4NGI1YjBhMDhkODJiNTAwMzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._HNGeJo3v_kSX2TvCkqf2pF4iGnQEY3Kv5eSa77SniY) See screenshot: I logged in as [electronics_user1](https://onto-deside.ilabt.imec.be/css11/electronics_user1/profile/card#me). The button 'request access' remains disabled and the field above it as well.

I've added some extra info pop-ups to the inputs and made the tree input reactive based on the state of the selected pods & their contents
![image](https://github.com/user-attachments/assets/65aa0d1c-23eb-491e-b0df-5049563128b8)
![image](https://github.com/user-attachments/assets/55996784-decd-4f25-9548-2bce6a539ec2)
![image](https://github.com/user-attachments/assets/69e43c46-6f9a-47a1-96e9-809f0552f245)

elsdvlee commented:
> > I have not looked at the code, but intended to test the functionality. Doing this, I ran into a first issue: it is unclear to me as a user how to request access. What is expected in the field WebID: the WebID to whom I want to sent an access request? What is expected in the field next to it? Where do I add the path of the file to which I request access? ![image](https://private-user-images.githubusercontent.com/93709574/389487308-f4864ceb-8c7e-4a37-b326-8690ae1cddcb.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI1NDcwODAsIm5iZiI6MTczMjU0Njc4MCwicGF0aCI6Ii85MzcwOTU3NC8zODk0ODczMDgtZjQ4NjRjZWItOGM3ZS00YTM3LWIzMjYtODY5MGFlMWNkZGNiLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDExMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMTI1VDE0NTk0MFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPThiMzcwNmUwZGI5MGMyZDBhMGUyYjQ4YmFlYzllNzE1MDE3MzhlODA3OGVmZGQ4NGI1YjBhMDhkODJiNTAwMzMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0._HNGeJo3v_kSX2TvCkqf2pF4iGnQEY3Kv5eSa77SniY) See screenshot: I logged in as [electronics_user1](https://onto-deside.ilabt.imec.be/css11/electronics_user1/profile/card#me). The button 'request access' remains disabled and the field above it as well.
> 
> I've added some extra info pop-ups to the inputs and made the tree input reactive based on the state of the selected pods & their contents ![image](https://private-user-images.githubusercontent.com/5462901/389586495-65aa0d1c-23eb-491e-b0df-5049563128b8.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI1NDkxOTMsIm5iZiI6MTczMjU0ODg5MywicGF0aCI6Ii81NDYyOTAxLzM4OTU4NjQ5NS02NWFhMGQxYy0yM2ViLTQ5MWUtYjBkZi01MDQ5NTYzMTI4YjgucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjVUMTUzNDUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9M2ViYWY5ZGFjZTQyZDViZjI4ZTkxY2QyZTEwYTU0ZTg4NjFjMDQ4Yzc2MDU0YWMzMzdlYmM3ZmU5YzAwNGEzMSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.wy142Iw3JuIs68OkGRwqfIpSOTNq60umLGoSEJGaEM4) ![image](https://private-user-images.githubusercontent.com/5462901/389586704-55996784-decd-4f25-9548-2bce6a539ec2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI1NDkxOTMsIm5iZiI6MTczMjU0ODg5MywicGF0aCI6Ii81NDYyOTAxLzM4OTU4NjcwNC01NTk5Njc4NC1kZWNkLTRmMjUtOTU0OC0yYmNlNmE1MzllYzIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjVUMTUzNDUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YTY3M2U1MWM3NTE3MDVlNGE3ZmZjZWRmZjUxYzNjMjMxODM1MWMwYjg0NWEzNzQxYjgxYmNkMDg0NGExMGE2MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.7M0mEx3Aa414jJO0C5awKbLX6_Kzn6WoHdaKoGa-xUE) ![image](https://private-user-images.githubusercontent.com/5462901/389587865-69e43c46-6f9a-47a1-96e9-809f0552f245.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzI1NDkxOTMsIm5iZiI6MTczMjU0ODg5MywicGF0aCI6Ii81NDYyOTAxLzM4OTU4Nzg2NS02OWU0M2M0Ni02ZjlhLTQ3YTEtOTZlOS04MDlmMDU1MmYyNDUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MTEyNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDExMjVUMTUzNDUzWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9YjI0Y2I5ODBlNTg5MTI4MjZkY2ZkZmNmYWQ1M2VkNjM0NzQwOGY4MzExMTk5ODFhZmIxYTVjMWM1Y2Y4NmU1MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.rff8Z6PdUMgr76OrQDjf20uschnc8B2ZEcYfmjutceI)

Thanks, this clarifies a lot.

NuttyShrimp commented:
Daarnaast ben ik ook nog bezig met venster toe te voegen waarin je kan selecteren welke permissies je precies wilt aanvragen. Momenteel is dat nog gehardcode op read permissies

elsdvlee commented on loama/src/router/index.ts:
```diff
@@ -14,9 +17,25 @@ const router = createRouter({
-            path: `/home/:filePath(.*)`,
-            name: 'home',
-            component: HomeView
+            path: "/",
```
Minor remark: the URIs http://localhost:5173/inbox and http://localhost:5173/request redirect to http://localhost:5173/home/. Can you add their paths?

NuttyShrimp replied:
Should be fixed, see 9718d982d888e0d81fd5ce1ce16edd210bd0a895

elsdvlee commented on loama/src/views/RequestView.vue:
```diff
@@ -0,0 +1,129 @@
+        await controller.value.AccessRequest().sendRequestNotification(session.info.webId!, checkedEntries)
+        toast.add({
+            severity: "success",
+            summary: "Access request(s) sended to user",
```
```suggestion
            summary: "Access request(s) sent to user",
```

elsdvlee commented on loama/src/components/inbox/RequestResponseMessage.vue:
```diff
@@ -0,0 +1,63 @@
+<template>
+    <div class="container">
+        <div><p class="prefix">Accepted:</p><PhCheckCircle v-if="message.isAccepted" weight="bold" /><PhXCircle v-else weight="bold" /></div>
+        <p><span class="prefix">To:</span> {{ message.target }}</p>
```
```suggestion
        <p><span class="prefix">Target:</span> {{ message.target }}</p>
```

elsdvlee commented on loama/src/components/inbox/RequestResponseMessage.vue:
```diff
@@ -0,0 +1,63 @@
+            <p v-for="label in aclLabels" :key="label">{{ label }}</p>
+        </div>
+        <div class="actions">
+            <LoButton @click="confirmMessage">Confirm</LoButton>
```
```suggestion
            <LoButton @click="confirmMessage">Mark as read</LoButton>
```

woutslabbinck commented:
First, I want to say I like how it works. 
Finally (given that both actors use loama), there is an easy way to request and grant access to other agents without having to deal with access control lists.

Currently, I have not checked the code yet as I have some high level questions:
- Is there a reason why public agents have `write` access to the inbox container (`<storage/public/loama/inbox>`)? Would `append` access be good enough?
	- `write` access could be abused by any actor to remove all resources there.
- Using the UI it is only possible to request `read` access. Would it be possible to also ask for the other types(`append`, `write`, `control`) as well?
- There is no visualization for pending access requests. How much work would it be to add that to e.g. the request access tab?
- For all the requests that you have received access, is there a list somewhere of that? If so, how much work would it be to visualize that?
- Maybe this last question is a bit stupid, but here we go. When an access request is handled, the notification in the inbox is removed. Right?
	- How much work would it be to keep a list somewhere of all historical requests/grants?

Another reason why I did not look into the code is that it seems to work as expected.
But, I'll take a look when these questions are answered as otherwise I might get too deep for this PR. I think that some of them might turn into new issues so we can merge this already to let other people use this version of loama.

NuttyShrimp committed:
feat(login): redirct to corrent path after login

NuttyShrimp committed:
fix(loama): label clarifications

NuttyShrimp committed:
fix(controller): only add append permission to inbox resource

NuttyShrimp committed:
feat: add permission options to requestable files

elsdvlee reviewed (approved):
Nice work! Tested and approved.

NuttyShrimp commented:
> First, I want to say I like how it works. Finally (given that both actors use loama), there is an easy way to request and grant access to other agents without having to deal with access control lists.
> 
> Currently, I have not checked the code yet as I have some high level questions:
> 
>     * Is there a reason why public agents have `write` access to the inbox container (`<storage/public/loama/inbox>`)? Would `append` access be good enough?
>       
>       * `write` access could be abused by any actor to remove all resources there.
> 
>     * Using the UI it is only possible to request `read` access. Would it be possible to also ask for the other types(`append`, `write`, `control`) as well?
> 
>     * There is no visualization for pending access requests. How much work would it be to add that to e.g. the request access tab?
> 
>     * For all the requests that you have received access, is there a list somewhere of that? If so, how much work would it be to visualize that?
> 
>     * Maybe this last question is a bit stupid, but here we go. When an access request is handled, the notification in the inbox is removed. Right?
>       
>       * How much work would it be to keep a list somewhere of all historical requests/grants?
> 
> 
> Another reason why I did not look into the code is that it seems to work as expected. But, I'll take a look when these questions are answered as otherwise I might get too deep for this PR. I think that some of them might turn into new issues so we can merge this already to let other people use this version of loama.

- I've updated the permission given to the resoources file to Append and everything seem to work
- As mentioned above, I started working on that when the PR was already opened so I could get early feedback, as of now this has been added to the PR
- It shouldn't take that long. It can be just another json file which can be cleaned with the accept/reject notifications
- If the previous request is fullfilled, we can use the pending request to store our needed data for this history

But I think it's better for the last two to be converted into seperate issues

woutslabbinck commented on controller/src/types/index.ts:
```diff
@@ -1,5 +1,10 @@
 import { url } from "loama-common";
 
+export interface Resources {
```
What is a Resource? Can you add some comments such that we know what is meant?

woutslabbinck commented on controller/src/types/index.ts:
```diff
@@ -39,5 +44,27 @@ export interface SubjectPermissions<T = BaseSubject<string>> {
     permissionsPerSubject: SubjectPermissions<T>[]
 }
+
+export interface ResourceAccessRequestNode {
```
I presume this is a recursive tree structure for a pod that checks where you can have access. Is that correct?
If so, can you document this at this interface?

I furthermore presume the children Record key is also an url, right?

NuttyShrimp replied:
I've added a comment to explain what this data structure should hold.
The first part of your comment is correct. But the children keys are path parts and not full urls to make tree visualizations easier

woutslabbinck commented on controller/src/types/modules.ts:
```diff
@@ -30,9 +32,37 @@ export interface IController<T extends Record<keyof T, BaseSubject<keyof T & str
+    getRequestableResources(containerUrl: string): Promise<ResourceAccessRequestNode>
+    canRequestAccessToResource(resourceUrl: string): Promise<boolean>
+    allowAccessRequest(resourceUrl: string): Promise<void>
+    disallowAccessRequest(resourceUrl: string): Promise<void>
```
Do this methods send HTTP requests to the acl files?

NuttyShrimp replied:
No, it sets if the resource is visible for access request

woutslabbinck commented on controller/src/types/modules.ts:
```diff
@@ -30,9 +32,37 @@ export interface IController<T extends Record<keyof T, BaseSubject<keyof T & str
+
+    loadAccessRequests(): Promise<AccessRequestMessage[]>;
+    loadRequestResponses(): Promise<RequestResponseMessage[]>;
+    removeRequest(messageUrl: string): Promise<void>;
```
What is the difference between `removeRequest(messageUrl: string): Promise<void>` and `disallowAccessRequest(resourceUrl: string): Promise<void>`

NuttyShrimp replied:
I've added explanation for both functions to the interfaces.
The first should remove the message from the inbox, the latter disable the resource from being requestable for another person

woutslabbinck reviewed (changes requested):
I focused on the interfaces and noticed that not all methods have documentation. It seems that `IStore`, `IController` and `IAccessRequest`  are crucial, yet only some methods in these interfaces are detailed. As a result, I am unsure what these methods specifically do. I've added remarks for the ones that are somewhat confusing to me. Could you add documentation describing their functions?

Another high-level aspect I do not understand is why the implementation of `AccessRequest` requires an `IController` and why the `IController` interface itself has a method `    AccessRequest(): IAccessRequest;`.
Could you elaborate how that works?

Finally, there is the granting `control` on access requests. 
We've added a safeguard (a warning clickbox) when you try to give access to other users when granting control access. However, when doing it via access requests, no message pop ups.
Could you add the same safeguard if the requested access includes control permission?

Other than that, it is ready to be merged.

Furthermore, I agree with your last message to make issues for the requested features and I will make those once this PR is merged.

NuttyShrimp committed:
feat(inbox): add confirmation for control permission

NuttyShrimp committed:
chore(types): add additional comments to interface functions

NuttyShrimp commented:
So, I've added explanation to most of the new functions in the interfaces.
Besides that, I've added a popup when accepting an access request that contains the "control" permission.
The `Controller` & `AccessRequest` referencing each other is so we just need to create a controller instance that contains everything.
The "problem" we're having here is that the "shared service" & "controller" logic is combined in 1 file, why the access request instance may look out of place.
The `AccessRequest` needs access to the controller so we can handle the permissions of access requests in the `AccessRequest` class
